L1Trigger/TrackerTFP/src/TrackQuality.cc

```diff
@@ -0,0 +1,258 @@
+    return lut;
+  }
+
+  //
+  template <class T>
+  int TrackQuality::toBin(const T& bins, double d) const {
+    int bin = 0;
+    for (; bin < (int)bins.size() - 1; bin++)
+      if (d < bins[bin + 1])
+        break;
+    return bin;
+  }
+
+  // Helper function to convert mvaPreSig to bin
+  int TrackQuality::toBinMVA(double mva) const {
+    static const array<double, numBinsMVA_> bins = mvaPreSigBins();
+    return toBin(bins, mva);
+  }
+
+  // Helper function to convert chi2B to bin
+  int TrackQuality::toBinChi2B(double chi2B) const {
+    static const array<double, numBinsChi2B_> bins = TTTrack_TrackWord::bendChi2Bins;
+    return toBin(bins, chi2B);
+  }
+
+  // Helper function to convert chi20 to bin
+  int TrackQuality::toBinChi20(double chi20) const {
+    static const array<double, numBinsChi20_> bins = TTTrack_TrackWord::chi2RPhiBins;
+    double chi2 = chi20 * chi2rphiConv_;
+    return toBin(bins, chi2);
+  }
+
+  // Helper function to convert chi21 to bin
+  int TrackQuality::toBinChi21(double chi21) const {
+    static const array<double, numBinsChi21_> bins = TTTrack_TrackWord::chi2RZBins;
+    double chi2 = chi21 * chi2rzConv_;
+    return toBin(bins, chi2);
+  }
+
+  TrackQuality::Track::Track(const FrameTrack& frameTrack, const StreamStub& streamStub, const TrackQuality* tq) {
+    static const DataFormats* df = tq->dataFormats();
+    static const Setup* setup = df->setup();
+    frames_.reserve(setup->tqNumChannel());
+    const TrackDR track(frameTrack, df);
+    double trackChi20(0.);
+    double trackChi21(0.);
+    TTBV hitPattern(0, streamStub.size());
+    for (int layer = 0; layer < (int)streamStub.size(); layer++) {
+      const FrameStub& frameStub = streamStub[layer];
+      if (frameStub.first.isNull())
+        continue;
+      const StubKF stub(frameStub, df);
+      hitPattern.set(layer);
+      const double m20 = tq->format(VariableTQ::m20).digi(pow(stub.phi(), 2));
+      const double m21 = tq->format(VariableTQ::m21).digi(pow(stub.z(), 2));
+      const double invV0 = tq->format(VariableTQ::invV0).digi(1. / pow(stub.dPhi(), 2));
+      const double invV1 = tq->format(VariableTQ::invV1).digi(1. / pow(stub.dZ(), 2));
+      const double stubChi20 = tq->format(VariableTQ::chi20).digi(m20 * invV0);
+      const double stubChi21 = tq->format(VariableTQ::chi21).digi(m21 * invV1);
+      trackChi20 += stubChi20;
+      trackChi21 += stubChi21;
+    }
+    if (trackChi20 > tq->format(VariableTQ::chi20).range())
+      trackChi20 = tq->format(VariableTQ::chi20).range() - 1e-11;
+    if (trackChi21 > tq->format(VariableTQ::chi21).range())
+      trackChi21 = tq->format(VariableTQ::chi21).range() - 1e-11;
+    // calc bdt inputs
+    const double cot = tq->scaleCot(df->format(Variable::cot, Process::dr).integer(track.cot()));
+    const double zT = tq->scaleZT(df->format(Variable::zT, Process::kf).integer(track.zT()));
+    const int nstub = hitPattern.count();
+    const int ninterior = nstub - 1;
```

Comment:
same as a couple of comments before, I don't want to change the behavior without knowing what the f/w expects. I would not use the HitPatternHelper, it is likely broken after this PR.

Reply:
The behavior has already been changed actually. Before this PR it was being calculated correctly, just in another way then I mentioned. So if you want to keep the firmware and software matching then it will need to be fixed. You can find the calculation before which Chris matched the firmware to [here](https://github.com/cms-L1TK/cmssw/blob/L1TK-dev-14_0_0_pre2/L1Trigger/TrackFindingTracklet/plugins/ProducerKFout.cc#L202-L231). I suggest using the simpler calculation which produces the same results and only builds off the hitpattern bits [here](https://github.com/cms-L1TK/cmssw/blob/L1TK-dev-14_0_0_pre2/L1Trigger/TrackTrigger/src/L1TrackQuality.cc#L39-L46).

Reply:
you are right, I overlooked a not. I fixed it.

Reply:
This calculation is still incorrect. The correct calculation using the TTBV class functions is ```hitPattern.count(hitPattern.plEncode(), hitPattern.pmEncode(), false)```. Can we also change the variable name to have "miss" in it? ninterior sounds like it should be counting the hits and not the misses. I suggest ```n_missint```.

Reply:
The calculations match the previous code, I don't want to change the behavior. I understand that it is likely buggy. But I can do the name change.

Reply:
Yes ok I do see that the calculation you have now has matches what was done before. I will open up an issue marking that both this and chi2rz and chi2rphi are incorrectly calculated so that we do not forget.